File: apps/master-maintenance-prj/node-src/src/audit.js
// 監査（誰が・いつ・何をしたか）。
//
// 案件の前書きで `audit` の問いにこう答えてある:
//   誰がいつ何を直したかを残す（削除と一括は必ず。監査表に残す）… 枠組みの外
//
// hatake は押されたことをアプリ側に渡すだけで、記録は持たない（誰がログインしているかも
// 知らない）。だからここで残す。

import { pool } from "./db.js";

export async function record({ user, action, target, key, detail }) {
  await pool.query(
    `insert into audit_log (user_id, action, target, target_key, detail)
     values ($1, $2, $3, $4, $5)`,
    [user, action, target, key ?? null, detail === undefined ? null : JSON.stringify(detail)],
  );
}

File: apps/order-entry-prj/java-src/src/main/java/io/example/order/web/Audit.java
package io.example.order.web;

import io.example.order.Db;
import org.springframework.stereotype.Component;

/**
 * 監査（誰が・いつ・何をしたか）。
 *
 * <p>案件の前書きで {@code audit} の問いにこう答えてある:
 * 「誰が・いつ・どの受注を直したかを残す」＝<b>枠組みの外</b>。
 * hatake は押されたことをアプリ側に渡すだけで、記録は持たない
 * （誰がログインしているかも知らない）。
 */
@Component
public class Audit {

    private final Db db;

    public Audit(Db db) {
        this.db = db;
    }

    public void record(User user, String action, String target, String key, String detail) {
        db.update(
                "insert into audit_log (user_id, action, target, target_key, detail)"
                        + " values (?, ?, ?, ?, ?)",
                user.userId(), action, target, key, detail);
    }
}
